### PR TITLE
[SPARK-49344][PS] Support `json_normalize` for Pandas API on Spark

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to support `json_normalize` for Pandas API on Spark.

### Why are the changes needed?

For Pandas feature parity: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.json_normalize.html

### Does this PR introduce _any_ user-facing change?

Introduce new API `ps.json_normalize`

```python
>>> data = [
...     {"id": 1, "name": "Alice", "address": {"city": "NYC", "zipcode": "10001"}},
...     {"id": 2, "name": "Bob", "address": {"city": "SF", "zipcode": "94105"}},
... ]
>>> ps.json_normalize(data)
   id   name address.city address.zipcode
0   1  Alice          NYC           10001
1   2    Bob           SF           94105
```

### How was this patch tested?

Added UTs

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by allisonwang-db — APPROVED** [review]
This is a very useful function!

**Comment by itholic** [comment]
Thanks @allisonwang-db for the review!

Also cc @ueshin @HyukjinKwon @xinrong-meng FYI

File: python/pyspark/pandas/namespace.py [hunk 1]
```diff
@@ -3687,6 +3689,80 @@ def read_orc(
     return psdf
 
 
+def json_normalize(
+    data: Union[Dict, List[Dict]],
+    sep: str = ".",
+) -> DataFrame:
+    """
+    Normalize semi-structured JSON data into a flat table.
+
```
Comment (zhengruifeng) [inline python/pyspark/pandas/namespace.py:3698]: nit
```suggestion
    .. versionadded:: 4.0.0

```

File: python/pyspark/pandas/namespace.py [hunk 2]
```diff
@@ -3687,6 +3689,81 @@ def read_orc(
     return psdf
 
 
+def json_normalize(
+    data: Union[Dict, List[Dict]],
+    sep: str = ".",
+) -> DataFrame:
+    """
+    Normalize semi-structured JSON data into a flat table.
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/namespace.py:3697]:
```suggestion
    Normalize semi-structured JSON data into a flat table.

```

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**